File: src/ui/clipboard.ts
// 复制文本到系统剪贴板
//
// 两条路径：
//   1. 平台命令（win: clip / mac: pbcopy / linux: wl-copy|xclip|xsel）—— 可靠、无长度限制；
//   2. OSC 52 —— 由终端负责写入剪贴板，本地没有任何剪贴板工具时仍可用
//      （Windows Terminal / iTerm2 / kitty / wezterm 均支持）。
import { spawn } from "node:child_process";

/** 按平台给出候选命令 */
function candidates(): Array<[string, string[]]> {
	if (process.platform === "win32") return [["clip", []]];
	if (process.platform === "darwin") return [["pbcopy", []]];
	return [
		["wl-copy", []],
		["xclip", ["-selection", "clipboard"]],
		["xsel", ["--clipboard", "--input"]],
	];
}

/** 逐个尝试平台命令写入剪贴板 */
function copyViaCommand(text: string): Promise<boolean> {
	const list = candidates();
	return new Promise<boolean>((resolve) => {
		const attempt = (index: number): void => {
			if (index >= list.length) {
				resolve(false);
				return;
			}
			const [cmd, args] = list[index];
			let child;
			try {
				child = spawn(cmd, args, { stdio: ["pipe", "ignore", "ignore"], windowsHide: true });
			} catch {
				attempt(index + 1);
				return;
			}
			let settled = false;
			const next = (ok: boolean): void => {
				if (settled) return;
				settled = true;
				if (ok) resolve(true);
				else attempt(index + 1);
			};
			child.on("error", () => next(false));
			child.on("close", (code) => next(code === 0));
			child.stdin.on("error", () => next(false));
			child.stdin.end(text, "utf8");
		};
		attempt(0);
	});
}

/** 复制文本；全部途径都失败时返回 false */
export async function copyToClipboard(text: string): Promise<boolean> {
	if (!text) return false;
	if (await copyViaCommand(text)) return true;
	if (process.stdout.isTTY) {
		process.stdout.write(`\u001b]52;c;${Buffer.from(text, "utf8").toString("base64")}\u0007`);
		return true;
	}
	return false;
}

File: rust/src/clipboard.rs
//! 往系统剪贴板写文本。
//!
//! 优先用本地工具，失败再退回 OSC 52 让终端自己处理：
//!   - WSL 下 `clip.exe` 直通 Windows 剪贴板，最省事
//!   - 桌面 Linux 常见 wl-copy（Wayland）/ xclip / xsel
//!   - OSC 52 在 Windows Terminal、iTerm2、kitty 等终端里都能用

use std::io::Write;
use std::process::{Command, Stdio};

fn tools() -> Vec<(&'static str, Vec<&'static str>)> {
    if std::path::Path::new("/mnt/c/Windows/System32/clip.exe").exists() {
        return vec![("clip.exe", vec![])];
    }
    match std::env::var("WAYLAND_DISPLAY") {
        Ok(_) => vec![("wl-copy", vec![]), ("xclip", vec!["-selection", "clipboard"])],
        Err(_) => vec![
            ("xclip", vec!["-selection", "clipboard"]),
            ("xsel", vec!["--clipboard", "--input"]),
            ("wl-copy", vec![]),
        ],
    }
}

fn pipe(tool: &str, args: &[&str], text: &str) -> bool {
    let Ok(mut child) = Command::new(tool)
        .args(args)
        .stdin(Stdio::piped())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
    else {
        return false;
    };
    if let Some(stdin) = child.stdin.as_mut() {
        if stdin.write_all(text.as_bytes()).is_err() {
            return false;
        }
    }
    child.wait().map(|s| s.success()).unwrap_or(false)
}

/// 复制文本；返回是否成功
pub fn copy(text: &str) -> bool {
    if text.is_empty() {
        return false;
    }
    for (tool, args) in tools() {
        if pipe(tool, &args, text) {
            return true;
        }
    }
    osc52(text)
}

/// OSC 52：把 base64 后的内容交给终端写进剪贴板
fn osc52(text: &str) -> bool {
    use base64::engine::general_purpose::STANDARD as B64;
    use base64::Engine as _;
    let encoded = B64.encode(text.as_bytes());
    // 有些终端对单条 OSC 长度有限制，过长时放弃
    if encoded.len() > 100_000 {
        return false;
    }
    let mut out = std::io::stdout();
    let _ = write!(out, "\u{1b}]52;c;{encoded}\u{7}");
    let _ = out.flush();
    true
}
